llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.h

```diff
@@ -174,24 +174,45 @@ class GCNTTIImpl final : public BasicTTIImplBase<GCNTTIImpl> {
   bool isAlwaysUniform(const Value *V) const;
 
   bool isValidAddrSpaceCast(unsigned FromAS, unsigned ToAS) const {
-    if (ToAS == AMDGPUAS::FLAT_ADDRESS) {
-      switch (FromAS) {
+    // Address space casts must cast between different address spaces.
+    if (FromAS == ToAS)
+      return false;
+
+    if (FromAS == AMDGPUAS::FLAT_ADDRESS ||
+        FromAS == AMDGPUAS::GLOBAL_ADDRESS ||
+        FromAS == AMDGPUAS::CONSTANT_ADDRESS ||
+        FromAS > AMDGPUAS::MAX_AMDGPU_ADDRESS) {
+      // Casting any 64-bit AS to another 64-bit AS or to a 32-bit AS is
+      // valid.
+      switch (ToAS) {
+      case AMDGPUAS::FLAT_ADDRESS:
       case AMDGPUAS::GLOBAL_ADDRESS:
-      case AMDGPUAS::CONSTANT_ADDRESS:
-      case AMDGPUAS::CONSTANT_ADDRESS_32BIT:
       case AMDGPUAS::LOCAL_ADDRESS:
+      case AMDGPUAS::CONSTANT_ADDRESS:
       case AMDGPUAS::PRIVATE_ADDRESS:
+      case AMDGPUAS::CONSTANT_ADDRESS_32BIT:
         return true;
       default:
         break;
       }
-      return false;
+      if (ToAS > AMDGPUAS::MAX_AMDGPU_ADDRESS)
+        return true;
+    } else if (FromAS == AMDGPUAS::LOCAL_ADDRESS ||
+               FromAS == AMDGPUAS::PRIVATE_ADDRESS ||
+               FromAS == AMDGPUAS::CONSTANT_ADDRESS_32BIT) {
+      // Casting from a 32-bit AS to any 64-bit AS is valid.
+      switch (ToAS) {
+      case AMDGPUAS::FLAT_ADDRESS:
+      case AMDGPUAS::GLOBAL_ADDRESS:
+      case AMDGPUAS::CONSTANT_ADDRESS:
+        return true;
+      default:
+        break;
```

Comment:
CONSTANT_ADDRESS_32BIT -> CONSTANT, GLOBAL, or FLAT is ok.

This incorrectly permits local->non-flat and private->non-flat

Reply:
Since addrspace casts are assumed to be reversible, should the inverse directions, i.e., non-flat -> (private or local), then also be forbidden?

Reply:
Assuming that is the case, I updated the PR with 0d29726202dffc47f000520ee648afece9bf3421 accordingly.

Reply:
Yes, p1 -> p3/p5 are also illegal